packages/aio-sk-lib-telemetry/source/core/instrumentation.ts

```diff
@@ -0,0 +1,408 @@
+export type RecursiveStringRecord = {
+  [key: string]: string | RecursiveStringRecord;
+};
+
+/** AsyncLocalStorage for helpers context. */
+const helpersStorage = new AsyncLocalStorage<InstrumentationHelpers>();
+
+/** Access helpers for the current instrumented operation. */
+export function getInstrumentationHelpers(): InstrumentationHelpers {
+  const context = helpersStorage.getStore();
+
+  if (!context) {
+    throw new Error(
+      "getInstrumentationHelpers can only be called from within an instrumented function",
+    );
+  }
+
+  return context;
+}
+
+/**
+ * Instruments a function.
+ * @param fn - The function to instrument.
+ * @param config - The configuration for the instrumentation.
+ * @returns A wrapped function with the same signature as the original function, but with telemetry instrumentation.
+ *
+ * @example
+ * ```ts
+ * const instrumentedFn = instrument(someFunction, {
+ *   // Optional configuration
+ *   spanConfig: {
+ *     spanName: "some-span",
+ *       attributes: {
+ *         "some-attribute": "some-value",
+ *       },
+ *   },
+ * });
+ */
+export function instrument<T extends AnyFunction>(
+  fn: T,
+  { spanConfig, hooks }: InstrumentationConfig<T> = {},
+): (...args: Parameters<T>) => ReturnType<T> {
+  const {
+    spanName = fn.name,
+    getBaseContext,
+    automaticSpanEvents = [],
+    ...spanOptions
+  } = spanConfig ?? {};
+
+  if (!spanName) {
+    throw new Error(
+      "Span name is required. Either provide a name or use a named function.",
+    );
+  }
+
+  const { onSuccess, onError } = hooks ?? {};
+  const autoEventSet = new Set(automaticSpanEvents);
+
+  /** Handles a (potentially) successful result within the given span. */
+  function handleResult(result: Awaited<ReturnType<T>>, span: Span) {
+    if (autoEventSet.has("success")) {
+      span.addEvent(`${fn.name ?? spanName}.success`, {
+        result: JSON.stringify(result),
+      });
+    }
+
+    // Temporal, handle pattern when we don't throw/error but instead return a "success" status.
+    if (result && typeof result === "object" && "success" in result) {
+      if (typeof result.success === "boolean" && result.success) {
+        onSuccess?.(result, span);
+        span.setStatus({ code: SpanStatusCode.OK });
+      } else {
+        onError?.(result, span);
+        span.setStatus({ code: SpanStatusCode.ERROR });
+      }
+    }
+
+    return result;
+  }
+
+  /** Handles an error result within the given span. */
+  function handleError(error: unknown, span: Span) {
+    if (autoEventSet.has("error")) {
+      span.addEvent(`${fn.name ?? spanName}.error`, {
+        error: JSON.stringify(error),
+      });
+    }
+
+    span.setStatus({ code: SpanStatusCode.ERROR });
+    const givenError = onError?.(error, span);
+
+    if (givenError) {
+      span.recordException(givenError);
+    } else if (error instanceof Error) {
+      span.recordException(error);
+    } else {
+      const exception = {
+        code: -1,
```

Comment:
Would it make sense to define a constant for this `-1` magic number?
By doing so, we could give it a meaningful name and explain what it is.